### PR TITLE
feat(forms): add submitted flag to NgForm

### PR DESCRIPTION
**Pull request by caitp**

### CODE REVIEW

**Comment by caitp** [comment]
questions:
  - when does the flag get reset
  - should it be a bound property
  - there has got to be a better way to test directive interactions with the DOM outside of E2E tests, but I don't see any examples in the tree.

File: modules/angular2/src/forms/directives/ng_form.ts [hunk 1]
```diff
@@ -51,12 +51,14 @@ const formDirectiveBinding =
   host: {
     '(submit)': 'onSubmit()',
   },
+  properties: ['submitted'],
```
Comment (caitp) [inline modules/angular2/src/forms/directives/ng_form.ts]: is this needed for the property to be accessed in expressions? It shouldn't be a bound property afaik

**Comment by juliemr** [comment]
Would something like https://github.com/angular/angular/blob/master/modules/angular2/test/debug/debug_element_spec.ts#L254 be the solution you're looking for? This shows triggering events from unit tests.

File: modules/examples/e2e_test/template_driven_forms/template_driven_forms_spec.ts [hunk 2]
```diff
@@ -18,4 +18,28 @@ describe('Template-Driven Forms', function() {
 
     expect(form.getInnerHtml()).toContain('is invalid credit card number');
   });
+
+
+  it('should set submitted flag', function() {
+    browser.get(URL);
+
+    var form = element.all(by.css('form')).first();
```
Comment (juliemr) [inline modules/examples/e2e_test/template_driven_forms/template_driven_forms_spec.ts]: You shouldn't need the `.first` calls here if IDs are unique - is this causing an issue?

Comment (caitp) [inline modules/examples/e2e_test/template_driven_forms/template_driven_forms_spec.ts]: no, it's just copy/pasted from the above spec for consistency

Comment (juliemr) [inline modules/examples/e2e_test/template_driven_forms/template_driven_forms_spec.ts]: I have no idea why the first spec is doing it that way. Seems much cleaner to me to do:

`var input = element(by.css('#creditCard'))`

**Comment by caitp** [comment]
TestComponentBuilder looks like it might work, i'll give that a try

File: modules/angular2/src/forms/directives/ng_form.ts (hunk 1 above)
Comment (vsavkin) [inline modules/angular2/src/forms/directives/ng_form.ts]: This is not needed. The `properties` array are the bindings of NgForm. It is required only when some other components sets submitted on NgForm using change detection.

**Comment by vsavkin** [comment]
>> when does the flag get reset

Not sure. When does Angular 1 reset similar flag?

>> should it be a bound property

I don't think so. You can use it in the template, but the value changes because of some event, not because of CD.

There are two implementations of form: NgForm and NgFormModel. Both have to have the submitted flag.

**Comment by caitp** [comment]
> Not sure. When does Angular 1 reset similar flag?

1.x resets the submitted flag during the `$setPristine()` method, which might be fine

**Comment by caitp** [comment]
This needs a bit more work to enable resetting the submitted state of the form, will be changnig duedate

**Comment by caitp** [comment]
so this might be ready for review?

File: modules/angular2/src/forms/directives/ng_form.ts (hunk 1 above)
Comment (vsavkin) [inline modules/angular2/src/forms/directives/ng_form.ts]: This needs to be removed.

File: modules/angular2/src/forms/directives/ng_form.ts [hunk 3]
```diff
@@ -124,10 +128,18 @@ export class NgForm extends ControlContainer implements Form {
   }
 
   onSubmit(): boolean {
+    this._submitted = true;
+    this._pristine = false;
```
Comment (vsavkin) [inline modules/angular2/src/forms/directives/ng_form.ts]: Why do we need this?

**Comment by vsavkin** [comment]
>> so this might be ready for review?

Please, update NgFormModel and address the comments.

**Comment by caitp** [comment]
> Please, update NgFormModel and address the comments.

@vsavkin can you please clarify this for me? Thanks.

**Comment by vsavkin** [comment]
>> @vsavkin can you please clarify this for me? Thanks.

Sure. 

* `properties: ['submitted']` is not required and should be removed.
* `_pristine` already exists on every control. I don't think we need it here.
* There are two form directives: NgForm and NgModelForm. You also need to update NgModelForm.

**Comment by tbosch** [comment]
@caitp ping, can you give this PR some love :-)

**Comment by caitp** [comment]
I'd love to, but the API has changed in ways that make no sense, and the documentation does not explain any of it.

**Comment by tbosch** [comment]
Let's close this for now. If we need it, we can redo it...

**Pull request closed without merging**